File: Lib/FlyCapture2/src/AsyncTriggerEx/AsyncTriggerEx.cpp
//=============================================================================
// Copyright © 2017 FLIR Integrated Imaging Solutions, Inc. All Rights Reserved.
//
// This software is the confidential and proprietary information of FLIR
// Integrated Imaging Solutions, Inc. ("Confidential Information"). You
// shall not disclose such Confidential Information and shall use it only in
// accordance with the terms of the license agreement you entered into
// with FLIR Integrated Imaging Solutions, Inc. (FLIR).
//
// FLIR MAKES NO REPRESENTATIONS OR WARRANTIES ABOUT THE SUITABILITY OF THE
// SOFTWARE, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
// PURPOSE, OR NON-INFRINGEMENT. FLIR SHALL NOT BE LIABLE FOR ANY DAMAGES
// SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING
// THIS SOFTWARE OR ITS DERIVATIVES.
//=============================================================================
//=============================================================================
// $Id: AsyncTriggerEx.cpp 316528 2017-02-22 00:03:53Z alin $
//=============================================================================

#include "stdafx.h"

#if defined(LINUX32) || defined(LINUX64)
#define LINUX
#endif

#ifdef LINUX
#include <time.h>
#include <unistd.h>
#endif

#include "FlyCapture2.h"
#include <iostream>
#include <sstream>

//
// Software trigger the camera instead of using an external hardware trigger
//
#define SOFTWARE_TRIGGER_CAMERA

using namespace FlyCapture2;
using namespace std;

void PrintBuildInfo()
{
    FC2Version fc2Version;
    Utilities::GetLibraryVersion(&fc2Version);

    ostringstream version;
    version << "FlyCapture2 library version: " << fc2Version.major << "."
            << fc2Version.minor << "." << fc2Version.type << "."
            << fc2Version.build;
    cout << version.str() << endl;

    ostringstream timeStamp;
    timeStamp << "Application build date: " << __DATE__ << " " << __TIME__;
    cout << timeStamp.str() << endl << endl;
}

void PrintCameraInfo(CameraInfo *pCamInfo)
{
    cout << endl;
    cout << "*** CAMERA INFORMATION ***" << endl;
    cout << "Serial number - " << pCamInfo->serialNumber << endl;
    cout << "Camera model - " << pCamInfo->modelName << endl;
    cout << "Camera vendor - " << pCamInfo->vendorName << endl;
    cout << "Sensor - " << pCamInfo->sensorInfo << endl;
    cout << "Resolution - " << pCamInfo->sensorResolution << endl;
    cout << "Firmware version - " << pCamInfo->firmwareVersion << endl;
    cout << "Firmware build time - " << pCamInfo->firmwareBuildTime << endl
         << endl;
}

void PrintError(Error error) { error.PrintErrorTrace(); }

bool CheckSoftwareTriggerPresence(Camera *pCam)
{
    const unsigned int k_triggerInq = 0x530;

    Error error;
    unsigned int regVal = 0;

    error = pCam->ReadRegister(k_triggerInq, &regVal);

    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return false;
    }

    if ((regVal & 0x10000) != 0x10000)
    {
        return false;
    }

    return true;
}

bool PollForTriggerReady(Camera *pCam)
{
    const unsigned int k_softwareTrigger = 0x62C;
    Error error;
    unsigned int regVal = 0;

    do
    {
        error = pCam->ReadRegister(k_softwareTrigger, &regVal);
        if (error != PGRERROR_OK)
        {
            PrintError(error);
            return false;
        }

    } while ((regVal >> 31) != 0);

    return true;
}

bool FireSoftwareTrigger(Camera *pCam)
{
    const unsigned int k_softwareTrigger = 0x62C;
    const unsigned int k_fireVal = 0x80000000;
    Error error;

    error = pCam->WriteRegister(k_softwareTrigger, k_fireVal);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return false;
    }

    return true;
}

int main(int /*argc*/, char ** /*argv*/)
{
    PrintBuildInfo();

    const int k_numImages = 10;

    Error error;

    BusManager busMgr;
    unsigned int numCameras;
    error = busMgr.GetNumOfCameras(&numCameras);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    cout << "Number of cameras detected: " << numCameras << endl;

    if (numCameras < 1)
    {
        cout << "Insufficient number of cameras... exiting" << endl;
        return -1;
    }

    PGRGuid guid;
    error = busMgr.GetCameraFromIndex(0, &guid);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    Camera cam;

    // Connect to a camera
    error = cam.Connect(&guid);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Power on the camera
    const unsigned int k_cameraPower = 0x610;
    const unsigned int k_powerVal = 0x80000000;
    error = cam.WriteRegister(k_cameraPower, k_powerVal);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    const unsigned int millisecondsToSleep = 100;
    unsigned int regVal = 0;
    unsigned int retries = 10;

    // Wait for camera to complete power-up
    do
    {
#if defined(_WIN32) || defined(_WIN64)
        Sleep(millisecondsToSleep);
#elif defined(LINUX)
        struct timespec nsDelay;
        nsDelay.tv_sec = 0;
        nsDelay.tv_nsec = (long)millisecondsToSleep * 1000000L;
        nanosleep(&nsDelay, NULL);
#endif
        error = cam.ReadRegister(k_cameraPower, &regVal);
        if (error == PGRERROR_TIMEOUT)
        {
            // ignore timeout errors, camera may not be responding to
            // register reads during power-up
        }
        else if (error != PGRERROR_OK)
        {
            PrintError(error);
            return -1;
        }

        retries--;
    } while ((regVal & k_powerVal) == 0 && retries > 0);

    // Check for timeout errors after retrying
    if (error == PGRERROR_TIMEOUT)
    {
        PrintError(error);
        return -1;
    }

    // Get the camera information
    CameraInfo camInfo;
    error = cam.GetCameraInfo(&camInfo);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    PrintCameraInfo(&camInfo);

#ifndef SOFTWARE_TRIGGER_CAMERA
    // Check for external trigger support
    TriggerModeInfo triggerModeInfo;
    error = cam.GetTriggerModeInfo(&triggerModeInfo);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    if (triggerModeInfo.present != true)
    {
        cout << "Camera does not support external trigger! Exiting..." << endl;
        return -1;
    }
#endif

    // Get current trigger settings
    TriggerMode triggerMode;
    error = cam.GetTriggerMode(&triggerMode);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Set camera to trigger mode 0
    triggerMode.onOff = true;
    triggerMode.mode = 0;
    triggerMode.parameter = 0;

#ifdef SOFTWARE_TRIGGER_CAMERA
    // A source of 7 means software trigger
    triggerMode.source = 7;
#else
    // Triggering the camera externally using source 0.
    triggerMode.source = 0;
#endif

    error = cam.SetTriggerMode(&triggerMode);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Poll to ensure camera is ready
    bool retVal = PollForTriggerReady(&cam);
    if (!retVal)
    {
        cout << endl;
        cout << "Error polling for trigger ready!" << endl;
        return -1;
    }

    // Get the camera configuration
    FC2Config config;
    error = cam.GetConfiguration(&config);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Set the grab timeout to 5 seconds
    config.grabTimeout = 5000;

    // Set the camera configuration
    error = cam.SetConfiguration(&config);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Camera is ready, start capturing images
    error = cam.StartCapture();
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

#ifdef SOFTWARE_TRIGGER_CAMERA
    if (!CheckSoftwareTriggerPresence(&cam))
    {
        cout << "SOFT_ASYNC_TRIGGER not implemented on this camera! Stopping "
                "application"
             << endl;
        return -1;
    }
#else
    cout << "Trigger the camera by sending a trigger pulse to GPIO"
         << triggerMode.source << endl;

#endif

    Image image;
    for (int imageCount = 0; imageCount < k_numImages; imageCount++)
    {

#ifdef SOFTWARE_TRIGGER_CAMERA
        // Check that the trigger is ready
        PollForTriggerReady(&cam);

        cout << "Press the Enter key to initiate a software trigger" << endl;
        cin.ignore();

        // Fire software trigger
        bool retVal = FireSoftwareTrigger(&cam);
        if (!retVal)
        {
            cout << endl;
            cout << "Error firing software trigger" << endl;
            return -1;
        }
#endif

        // Grab image
        error = cam.RetrieveBuffer(&image);
        if (error != PGRERROR_OK)
        {
            PrintError(error);
            break;
        }
        cout << "." << endl;
    }

    // Turn trigger mode off.
    triggerMode.onOff = false;
    error = cam.SetTriggerMode(&triggerMode);
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }
    cout << endl;
    cout << "Finished grabbing images" << endl;

    // Stop capturing images
    error = cam.StopCapture();
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    // Disconnect the camera
    error = cam.Disconnect();
    if (error != PGRERROR_OK)
    {
        PrintError(error);
        return -1;
    }

    cout << "Done! Press Enter to exit..." << endl;
    cin.ignore();

    return 0;
}

File: Lib/FlyCapture2/src/AsyncTriggerEx_VB.Net/AsyncTriggerEx.vb
'=============================================================================
' Copyright © 2017 FLIR Integrated Imaging Solutions, Inc. All Rights Reserved.
'
' This software Is the confidential And proprietary information of FLIR
' Integrated Imaging Solutions, Inc. ("Confidential Information"). You
' shall Not disclose such Confidential Information And shall use it only in
' accordance with the terms of the license agreement you entered into
' with FLIR Integrated Imaging Solutions, Inc. (FLIR).
'
' FLIR MAKES NO REPRESENTATIONS Or WARRANTIES ABOUT THE SUITABILITY OF THE
' SOFTWARE, EITHER EXPRESSED Or IMPLIED, INCLUDING, BUT Not LIMITED TO, THE
' IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
' PURPOSE, Or NON-INFRINGEMENT. FLIR SHALL Not BE LIABLE FOR ANY DAMAGES
' SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING Or DISTRIBUTING
' THIS SOFTWARE Or ITS DERIVATIVES.
'=============================================================================
' =============================================================================
' $Id: AsyncTriggerEx.vb 317565 2017-03-01 21:05:37Z alin $
' =============================================================================

Imports System
Imports System.Text
Imports FlyCapture2Managed

Namespace AsyncTriggerEx_VB
    Class Program

        Shared Sub PrintBuildInfo()

            Dim version As FC2Version = ManagedUtilities.libraryVersion
            Dim newStr As StringBuilder = New StringBuilder()
            newStr.AppendFormat("FlyCapture2 library version: {0}.{1}.{2}.{3}" & vbNewLine, _
                                version.major, version.minor, version.type, version.build)
            Console.WriteLine(newStr)

        End Sub


        Shared Sub PrintCameraInfo(ByVal camInfo As CameraInfo)

            Dim newStr As StringBuilder = New StringBuilder()
            newStr.Append(vbNewLine & "*** CAMERA INFORMATION ***" & vbNewLine)
            newStr.AppendFormat("Serial number - {0}" & vbNewLine, camInfo.serialNumber)
            newStr.AppendFormat("Camera model - {0}" & vbNewLine, camInfo.modelName)
            newStr.AppendFormat("Camera vendor - {0}" & vbNewLine, camInfo.vendorName)
            newStr.AppendFormat("Sensor - {0}" & vbNewLine, camInfo.sensorInfo)
            newStr.AppendFormat("Resolution - {0}" & vbNewLine, camInfo.sensorResolution)

            Console.WriteLine(newStr)

        End Sub

        Shared Function CheckSoftwareTriggerPresence(ByVal cam As ManagedCamera) As Boolean

            Const TriggerInquiry As UInteger = &H530
            Dim regVal As UInteger = cam.ReadRegister(TriggerInquiry)

            If ((regVal And &H10000) <> &H10000) Then
                Return False
            End If

            Return True

        End Function

        Shared Function PollForTriggerReady(ByVal cam As ManagedCamera) As Boolean

            Const SoftwareTrigger As UInteger = &H62C
            Dim SoftwareTriggerValue As UInteger = 0

            Do
                SoftwareTriggerValue = cam.ReadRegister(SoftwareTrigger)
            Loop While ((SoftwareTriggerValue >> 31) <> 0)

            Return True

        End Function

        Shared Function FireSoftwareTrigger(ByVal cam As ManagedCamera) As Boolean

            Const SoftwareTrigger As UInteger = &H62C
            Const SoftwareTriggerFireValue As UInt32 = &H80000000UI

            cam.WriteRegister(SoftwareTrigger, SoftwareTriggerFireValue)
            Return True

        End Function

        Shared Sub Main()

            PrintBuildInfo()

            Const NumImages As Integer = 10
            Dim useSoftwareTrigger As Boolean = True

            Dim busMgr As ManagedBusManager = New ManagedBusManager()
            Dim numCameras As UInteger = busMgr.GetNumOfCameras()

            Console.WriteLine("Number of cameras detected: {0}", numCameras)

            If numCameras < 1 Then
                Console.WriteLine("No cameras detected!")
                Console.WriteLine("Press enter to exit...")
                Console.ReadLine()
                Return
            End If

            Dim guid As ManagedPGRGuid = busMgr.GetCameraFromIndex(0)

            Dim cam As ManagedCamera = New ManagedCamera()

            cam.Connect(guid)

            ' Power on the camera
            Const CameraPower As UInteger = &H610
            Const CameraPowerValue As UInt32 = &H80000000UI
            cam.WriteRegister(CameraPower, CameraPowerValue)

            ' Wait for camera to complete power-up
            Const MillisecondsToSleep = 100
            Dim regVal As UInteger = 0

            Do While ((regVal And CameraPowerValue) = 0)
                System.Threading.Thread.Sleep(MillisecondsToSleep)
                regVal = cam.ReadRegister(CameraPower)
            Loop

            ' Get the camera information
            Dim camInfo As CameraInfo = cam.GetCameraInfo()

            PrintCameraInfo(camInfo)

            If (Not useSoftwareTrigger) Then

                ' Check for external trigger support
                Dim triggerModeInfo As TriggerModeInfo = cam.GetTriggerModeInfo()

                If (Not triggerModeInfo.present) Then
                    Console.WriteLine("Camera does not support external trigger!")
                    Console.WriteLine("Press enter to exit...")
                    Console.ReadLine()
                    Return
                End If
            End If

            ' Get current trigger settings
            Dim triggerMode As TriggerMode = cam.GetTriggerMode()

            ' Set camera to trigger mode 0
            ' A source of 7 means software trigger
            triggerMode.onOff = True
            triggerMode.mode = 0
            triggerMode.parameter = 0

            If (useSoftwareTrigger) Then
                ' A source of 7 means software trigger
                triggerMode.source = 7
            Else
                ' Triggering the camera externally using source 0.
                triggerMode.source = 0
            End If


            ' Set the trigger mode
            cam.SetTriggerMode(triggerMode)

            ' Poll to ensure camera is ready
            Dim retVal As Boolean = PollForTriggerReady(cam)

            If (Not retVal) Then
                Return
            End If

            ' Get the camera configuration
            Dim config As FC2Config = cam.GetConfiguration()

            ' Set the grab timeout to 5 seconds
            config.grabTimeout = 5000

            ' Set the camera configuration
            cam.SetConfiguration(config)

            ' Camera is ready, start capturing images
            cam.StartCapture()

            If (useSoftwareTrigger) Then
                If (Not CheckSoftwareTriggerPresence(cam)) Then
                    Console.WriteLine("SOFT_ASYNC_TRIGGER not implemented on this camera!")
                    Console.WriteLine("Press enter to exit...")
                    Console.ReadLine()
                    Return
                End If
            Else
                Console.WriteLine("Trigger the camera by sending a trigger pulse to GPIO%d." & vbNewLine, _
                triggerMode.source)
            End If

            Dim rawImage As ManagedImage = New ManagedImage()
            For iImageCount As Integer = 0 To (NumImages - 1)

                If (useSoftwareTrigger) Then

                    ' Check that the trigger is ready
                    retVal = PollForTriggerReady(cam)

                    Console.WriteLine("Press the Enter key to initiate a software trigger." & vbNewLine)
                    Console.ReadLine()

                    ' Fire software trigger
                    retVal = FireSoftwareTrigger(cam)
                    If (Not retVal) Then
                        Console.WriteLine("Error firing software trigger!")
                        Console.WriteLine("Press enter to exit...")
                        Console.ReadLine()
                        Return
                    End If
                End If

                Try
                    ' Retrieve an image
                    cam.RetrieveBuffer(rawImage)
                Catch ex As FC2Exception
                    Console.WriteLine("Error retrieving buffer : {0}", ex.Message)
                    Continue For
                End Try

                Console.WriteLine("." & vbNewLine)
            Next

            Console.WriteLine("Finished grabbing images")

            ' Stop capturing images
            cam.StopCapture()

            ' Turn off trigger mode
            triggerMode.onOff = False
            cam.SetTriggerMode(triggerMode)

            ' Disconnect the camera
            cam.Disconnect()

            Console.WriteLine("Done! Press enter to exit...")
            Console.ReadLine()
        End Sub

    End Class
End Namespace
